File: haskell/ac11.hs
import Data.Array
import Data.List

distances galaxies emptyRows emptyColumns factor =
    sum [ sum [ abs (x1 - x2) + abs (y1 - y2) +
                factor * (abs (emptyRows ! y1 - emptyRows ! y2) +
                         abs (emptyColumns ! x1 - emptyColumns ! x2))
              | (x2, y2) <- ts ]
          | ((x1, y1):ts) <- tails galaxies ]

main = do
    contents <- getContents
    let grid = map (map (== '#')) (lines contents)
    let first:_ = grid
    let width = length first
    let emptyRows =
          listArray (0, length grid)
          (scanl1 (+) (map (fromEnum . all not) grid))
    let emptyColumns =
          listArray (0, width)
          (scanl1 (+) (map (fromEnum . all not) (transpose grid)))
    let galaxies = [
          (x, y) | (y, row) <- zip [0..] grid, (x, True) <- zip [0..] row]
    let resultPart1 = distances galaxies emptyRows emptyColumns 1
    putStrLn("Part 1: " ++ show resultPart1)
    let resultPart2 = distances galaxies emptyRows emptyColumns (1000000 - 1)
    putStrLn("Part 2: " ++ show resultPart2)

File: rust/src/bin/ac11.rs
fn partial_sums(it: impl Iterator<Item = u64>) -> impl Iterator<Item = u64> {
    it.scan(0, |state, x| {
	*state += x;
	Some(*state)
    })
}

fn distances(
    galaxies: &Vec<(usize, usize)>,
    empty_rows: &Vec<u64>,
    empty_columns: &Vec<u64>,
    factor: u64
) -> u64 {
    galaxies.iter().enumerate().map(|(i, &(x1, y1))|
        (&galaxies[i+1..]).iter().map(|&(x2, y2)|
            (x1 as i64 - x2 as i64).abs() as u64 +
	    (y1 as i64 - y2 as i64).abs() as u64 + 
            (
		(empty_rows[y1] as i64 - empty_rows[y2] as i64).abs() +
		(empty_columns[x1] as i64 - empty_columns[x2] as i64).abs()
	    ) as u64 * factor
	  ).sum::<u64>()
    ).sum()
}

fn main() {
    let lines = std::io::stdin().lines();
    let grid: Vec<Vec<_>> =
	lines.map(|line| line.unwrap().chars().map(|c| c == '#').
        collect()).collect();
    let width = grid.first().unwrap().len();
    let empty_rows: Vec<_> =
	partial_sums(
	    grid.iter().map(|line| line.iter().all(std::ops::Not::not) as u64)
	).collect();
    let empty_columns: Vec<_> =
	partial_sums(
	    (0..width).map(|index|
                grid.iter().all(|line| !line[index]) as u64
	    )
	).collect();
    let galaxies: Vec<_> =
	grid.iter().enumerate().flat_map(|(y, line)|
          line.iter().enumerate().filter_map(move |(x, &galaxy)|
            galaxy.then_some((x, y))
	  )).collect();
    let result_part1 = distances(&galaxies, &empty_rows, &empty_columns, 1);
    let result_part2 =
	distances(&galaxies, &empty_rows, &empty_columns, 1000000 - 1);
    println!("Part 1: {result_part1}");
    println!("Part 2: {result_part2}");
}
